### PR TITLE
Make UnstructuredContent return contents without mutating the source

### LINKED ISSUE #56316: runtime.Unstructured.UnstructuredContent() is easy to misuse
**What happened**:
In #51526 I introduced an optimization - do a deep copy instead of to and from JSON roundtrip to convert anything that implements `runtime.Unstructured`. I then discovered that the method that is used there `UnstructuredContent()` in both `Unstructured` and `UnstructuredList` may mutate the original object.
https://github.com/kubernetes/kubernetes/blob/160cf97fa426dcf355357a81e7824840afd52bcc/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L83-L88
https://github.com/kubernetes/kubernetes/blob/160cf97fa426dcf355357a81e7824840afd52bcc/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go#L55-L72
This is problematic because previously (before #51526) there was no mutation and because this is unexpected and may lead to data races - it is bad behaviour to mutate original object when you just want a copy of it.
**This issue was fixed by #55871**.

However the underlying issue is not fixed here - `UnstructuredContent()` is brittle and dangerous. Method name does not imply that it mutates data when you call it. And godoc does not mention that either:
https://github.com/kubernetes/kubernetes/blob/e7e313d6d23f4e4619944049a265270353c19cf8/staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go#L233-L249
Something needs to be done about it IMO.

This method is misused in these places (mutation side effect is not intended):
https://github.com/kubernetes/kubernetes/blob/8442a118ea91c1ccb6a4b2d1505359732608b5cf/pkg/printers/jsonpath.go#L117-L152
https://github.com/kubernetes/kubernetes/blob/091a9dfc35a10637c464b7f5e1e2ec2d25d7067f/pkg/printers/customcolumn.go#L203-L242
https://github.com/kubernetes/kubernetes/blob/d2a62fd42234a96cbab2dbcf402c168c59b41784/pkg/printers/humanreadable.go#L685-L742

Also `UnstructuredContent()` implementation in `UnstructuredList` does not implement the behaviour required by godoc in `runtime.Unstructured`.

**What you expected to happen**:
There should be a method that returns the current unstructured contents without mutation and perhaps not providing the write-through effect.

/kind bug
/sig api-machinery

### PR DESCRIPTION
**Pull request by ash2k**
**What this PR does / why we need it**:
This PR solves the issues described in #56316

Before this change:

- A call to `UnstructuredContent()` potentially modified `Object`
- The values returned by `UnstructuredContent()` could be manipulated to modify the value in `Object`. Going through the history it looks like this behavior was added before the addition of `SetUnstructuredContent()`. IMO it makes more sense now to use `SetUnstructuredContent()` or make changes to the exposed `Object` property
- `UnstructuredList` did not implement the behavior described in the godoc. The godoc stated that the value returned should be mutable, but if u.Object == nil the map returned had no effect on Object

With this PR I'm proposing `UnstructuredContent()` returns the data without providing the contract of a mutable map. It also ensures all implementations of the `Unstructured` interface abide by the doc

**Which issue(s) this PR fixes**:
Fixes #56316

**Special notes for your reviewer**:
This PR continues work started in #57713.

**Release note**:
```release-note
NONE
```
/kind bug
/sig api-machinery
/cc sttts deads2k

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go [hunk 1]
```diff
@@ -53,19 +53,21 @@ func (u *UnstructuredList) EachListItem(fn func(runtime.Object) error) error {
...
-// returned object will affect items in the underlying Items field, but changing
-// the "items" slice itself will have no effect.
-// TODO: expose SetUnstructuredContent on runtime.Unstructured that allows
-// items to be changed.
+// the Object field. Items always overwrites overlay.
 func (u *UnstructuredList) UnstructuredContent() map[string]interface{} {
-	out := u.Object
-	if out == nil {
-		out = make(map[string]interface{})
+	out := make(map[string]interface{}, len(u.Object)+1)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go]: why +1?

**Comment by sttts** [comment]
Could we make `Unstructured/UnstructuredList.Object` private?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go (hunk 1 above)
Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go]: The contents may not have `items`.

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go]: Ok, so the case in line 62 should not happen?

**Comment by sttts** [comment]
A constructor sounds good.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go (hunk 1 above)
Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go]: `u.Object` may or may not contain `items`. We can remove `if` in the loop because the key will be overwritten anyway. It made sense to have it before (in the previous PR) because a deep copy was made here and skipping `items` reduced the unnecessary work. WDYT?

**Comment by ash2k** [comment]
@sttts can we merge this one first and then I'll submit another one with the constructor change? It is going to be a big PR and I think it makes sense to keep the changes separate. Also this is a bugfix and a refactoring - should be separate.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go (hunk 1 above)
Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go]: I've removed the `if`.

File: pkg/printers/internalversion/describe_test.go [hunk 2]
```diff
@@ -2133,14 +2133,14 @@ URL:	http://localhost
 				"name":              "MyName",
 				"namespace":         "MyNamespace",
 				"creationTimestamp": "2017-04-01T00:00:00Z",
-				"resourceVersion":   123,
+				"resourceVersion":   int64(123),
```
Comment (sttts) [inline pkg/printers/internalversion/describe_test.go]: this is unrelated. Can you move it to another PR? (I cannot approve this).

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNilUnstructuredContent(t *testing.T) {
+	var u Unstructured
+	content := u.UnstructuredContent()
+	expContent := make(map[string]interface{})
+	assert.EqualValues(t, expContent, content)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: Can you check that the actual object stays the same? Make a deep copy before and then compare afterwards.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go]: 2018

**Comment by sttts** [comment]
> can we merge this one first and then I'll submit another one with the constructor change?

yes, makes sense.

File: pkg/printers/internalversion/describe_test.go (hunk 2 above)
Comment (ash2k) [reply pkg/printers/internalversion/describe_test.go]: Most likely it was causing a failure because `123` literal has an `int` type by default but deep copy only likes `float64` and `int64`. But this fix is no longer needed because we are not making a deep copy.

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go [hunk 5]
```diff
@@ -446,20 +446,31 @@ func DeepCopyJSON(x map[string]interface{}) map[string]interface{} {
 // DeepCopyJSONValue deep copies the passed value, assuming it is a valid JSON representation i.e. only contains
 // types produced by json.Unmarshal().
 func DeepCopyJSONValue(x interface{}) interface{} {
+	if x == nil {
+		return nil
+	}
 	switch x := x.(type) {
 	case map[string]interface{}:
+		if x == nil {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: does this really exist?

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Yes, without this change (and below for completeness) the newly added check in the test (deep copy and compare) was failing. https://play.golang.org/p/KASoa2DeTnp

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: This will change the output - this code used to produce non-nil maps and slices for typed nils but now will produce typed nils. JSON encoding will change - typed nils as `null` in JSON (https://play.golang.org/p/Uz2ajLEQx9A).

I think this change is a good one - it makes the deep copy more precise.

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: We should never see typed nils somewhere deep inside IMO. On the toplevel maybe.

Compare https://play.golang.org/p/R45uKKqDsm6

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Yes, we should not see them, but they exist. I'm fine with removing this fix from this PR if that's what you think is best. However I think that way deep copy is more correct. It just mirrors the input precisely. If there are no typed nils, we don't introduce them here. If they create a problem somewhere then the place that "spawns" them should be fixed. Or, even better, a place that does not handle them well could be made more robust.

Also, I'm not sure what exactly you wanted me to compare? typed and untyped nils are both marshaled into `null`, yes :)

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Have we ever seen this "cannot deep copy" panic? You say that they show up. Do you know where?

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Panic? Current code does not cause any panics AFAIK. Are you talking about failing test without these changes? It does not panic, it just fails because objects are not equal. They are not equal because the deep copied `Object` field ("expected" in the test) is an empty map and the original is `nil.

```go
func TestNilUnstructuredContent(t *testing.T) {
	var u Unstructured
	uCopy := u.DeepCopy()
	content := u.UnstructuredContent()
	expContent := make(map[string]interface{})
	assert.EqualValues(t, expContent, content)
	assert.Equal(t, uCopy, &u)
}
```
```
--- FAIL: TestNilUnstructuredContent (0.00s)
	Error Trace:	unstructured_test.go:31
	Error:      	Not equal: 
	            	expected: &unstructured.Unstructured{Object:map[string]interface {}{}}
	            	actual: &unstructured.Unstructured{Object:map[string]interface {}(nil)}
	            	
	            	Diff:
	            	--- Expected
	            	+++ Actual
	            	@@ -1,4 +1,3 @@
	            	 (*unstructured.Unstructured)({
	            	- Object: (map[string]interface {}) {
	            	- }
	            	+ Object: (map[string]interface {}) <nil>
	            	 })
```

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go [hunk 6]
```diff
@@ -447,20 +446,31 @@ func DeepCopyJSON(x map[string]interface{}) map[string]interface{} {
...
+			// Typed nil - an interface{} that contains a type []interface{} with a value of nil
+			return x
+		}
 		clone := make([]interface{}, len(x))
 		for i, v := range x {
 			clone[i] = DeepCopyJSONValue(v)
 		}
 		return clone
-	case string, int64, bool, float64, nil, encodingjson.Number:
+	case string, int64, bool, float64, encodingjson.Number:
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: why not keep the nil and remove the if clause above?

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Don't feel strong about it. Was just curious.

**Comment by sttts** [comment]
the verify job is broken: https://github.com/kubernetes/kubernetes/pull/62162

/lgtm
/approve

**Comment by wenjiaswe** [comment]
cc @mbohlool
